### PR TITLE
refactor: remove electron.asar and embed JS in binary

### PR DESCRIPTION
**Pull request by MarshallOfSound**
#### Description of Change
Does what it says on the tin, some more performance gains post-bundling.  This embeds the three `init.js` files in the binary with js2c and includes some changes in node (https://github.com/electron/node/pull/101) that allow us to use the built-in modules as entrypoints.

#### Release Notes

Notes: The `electron.asar` file no longer exists, if you have packaging scripts that depend on it's existence you should update them.

### CODE REVIEW

**Comment by miniak** [comment]
@MarshallOfSound how does this change affect debugging?

**Comment by MarshallOfSound** [comment]
@miniak Just means that instead of using `--inspect-brk` to debug Electron internals you have to use `--inspect-brk-node`.  I have a future plan of adding `--inspect-brk-electron` but for you can debug things just fine 👍

File: lib/browser/init.ts [hunk 1]
```diff
@@ -106,7 +106,7 @@ if (process.resourcesPath) {
   for (packagePath of searchPaths) {
     try {
       packagePath = path.join(process.resourcesPath, packagePath)
-      packageJson = __non_webpack_require__(path.join(packagePath, 'package.json')) // eslint-disable-line
+      packageJson = Module._load(path.join(packagePath, 'package.json'))
```
Comment (nornagon) [inline lib/browser/init.ts]: this seems unrelated to me...?

File: lib/renderer/init.ts [hunk 2]
```diff
@@ -203,7 +205,7 @@ for (const preloadScript of preloadScripts) {
     if (!isParentDir(getAppPath(), fs.realpathSync(preloadScript))) {
       throw new Error('Preload scripts outside of app path are not allowed')
     }
-    __non_webpack_require__(preloadScript)  // eslint-disable-line
+    Module._load(preloadScript)  // eslint-disable-line
```
Comment (nornagon) [inline lib/renderer/init.ts]: what lint are we disabling here?

File: lib/worker/init.js [hunk 3]
```diff
@@ -27,8 +28,9 @@ if (self.location.protocol === 'file:') {
   global.module.filename = global.__filename
 
   // Also search for module under the html file.
-  global.module.paths = global.module.paths.concat(Module._nodeModulePaths(global.__dirname))
+  global.module.paths = Module._nodeModulePaths(global.__dirname)
 } else {
-  global.__filename = __filename
-  global.__dirname = __dirname
+  // For backwards compatibility we fake these two paths here
```
Comment (nornagon) [inline lib/worker/init.js]: what depends on this?

File: lib/browser/init.ts (hunk 1 above)
Comment (MarshallOfSound) [reply lib/browser/init.ts]: `__non_webpack_require__` maps to the global node provided require function, now we are an install module that require function is `NativeModule.require` which does not load non-internal code.  We need to explicitly load non-internal code so we do what node does internally and directly call `Module._load`

File: lib/renderer/init.ts (hunk 2 above)
Comment (MarshallOfSound) [reply lib/renderer/init.ts]: Can't remember, might be a leftover from the `__non*`

File: lib/worker/init.js (hunk 3 above)
Comment (MarshallOfSound) [reply lib/worker/init.js]: User code, `__dirname` and `__filename` are globals in node land that are expected to exist and be valid.  When the internal `init` doesn't exist on disk node doesn't set these values so here I fake them to their old values.  In all likelihood we will have to keep this logic forever 👍

Comment (nornagon) [reply lib/worker/init.js]: wouldn't those variables be in relation to the user's files though, not the asar?
